torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -6,16 +6,171 @@ namespace rpc {
+    std::unique_lock<std::mutex> lock(rpcRetryMutex_);
 
-RpcAgent::~RpcAgent() = default;
+    // We must continue sleeping as long as the RPC Agent is running and when
+    // either the Retry Map is empty, or when the Retry Map's earliest expiring
+    // RPC is set to be retried in the future.
+    if (rpcRetryMap_.empty()) {
+      rpcRetryMapCV_.wait(lock, [&, this] {
+        return !(rpcAgentRunning_.load() && rpcRetryMap_.empty());
+      });
+    }
+
+    steady_clock_time_point earliestTimeout;
+
+    if (!rpcRetryMap_.empty()) {
+      earliestTimeout = rpcRetryMap_.begin()->first;
+      rpcRetryMapCV_.wait_until(lock, earliestTimeout, [&, this] {
+        return !(rpcAgentRunning_.load());
+      });
+    }
+
+    if (!rpcAgentRunning_.load()) {
+      return;
+    }
+
+    // Updating these since something may have been added to the map while this
+    // thread was sleeping.
+    earliestTimeout = rpcRetryMap_.begin()->first;
+    auto& earliestRpcList = rpcRetryMap_.begin()->second;
+
+    // We iterate through all the RPC's set to be retried at the current
+    // timepoint, resend those RPC's, and add a callback that either schedules
+    // the RPC for a future retry or marks it with success/error depending on
+    // the outcome of the current send. Then, we clean up the rpcRetryMap_.
+    for (auto it = earliestRpcList.begin(); it != earliestRpcList.end();
+         /* no increment */) {
+      auto& earliestRpc = *it;
+      steady_clock_time_point newTime =
+          computeNewRpcRetryTime(earliestRpc->retryCount_);
+      auto fm = send(earliestRpc->to_, std::move(earliestRpc->message_));
+
+      fm->addCallback([this, earliestRpc, newTime](
+                          const rpc::Message& message,
+                          const c10::optional<utils::FutureError>& futErr) {
+        if (futErr) {
+          LOG(INFO) << "Send try "
+                    << std::to_string(earliestRpc->retryCount_ + 1)
+                    << " failed\n";
+          earliestRpc->retryCount_ += 1;
+          if (earliestRpc->retryCount_ < earliestRpc->maxRetries_) {
+            // If the previous future completed with an error and we haven't
+            // completed maxRetries send attempts, we move the earliestRpc
+            // struct to a new time point in the retry map (effectively
+            // scheduling it for a future retry.)
+            {
+              std::lock_guard<std::mutex> retryMapLock(rpcRetryMutex_);
+              rpcRetryMap_[newTime].emplace(std::move(earliestRpc));
+            }
+          } else {
+            // We have completed maxRetries send attempts. We're now marking
+            // the future with an error.
+            std::string errorMessage;
+            errorMessage
+                .append(
+                    "The RPC has not succeeded after the specified number of max retries (")
+                .append(std::to_string(earliestRpc->maxRetries_))
+                .append(").\n");
```

Comment:
Use `c10::str(` instead to build the string here. Also, note that you don't need the trailing \n